src/providers/open-ai-base/index.ts

```diff
@@ -0,0 +1,378 @@
+  T extends OpenAICompleteResponse | ErrorResponse,
+>(
+  provider: string,
+  customTransformer?: CustomTransformer<OpenAICompleteResponse, T>
+) => {
+  const transformer: (
+    response: T | ErrorResponse,
+    responseStatus: number
+  ) => T | ErrorResponse = (response, responseStatus) => {
+    if (responseStatus !== 200 && 'error' in response) {
+      const errorResponse = OpenAIErrorResponseTransform(
+        response,
+        provider ?? OPEN_AI
+      );
+      if (customTransformer) {
+        return customTransformer(errorResponse, true);
+      }
+    }
+
+    if (customTransformer) {
+      return customTransformer(response as T);
+    }
+
+    Object.defineProperty(response, 'provider', {
+      value: provider,
+      enumerable: true,
+    });
+
+    return response;
+  };
+
+  return transformer;
+};
+
+const ChatCompleteResponseTransformer = <
+  T extends OpenAIChatCompleteResponse | ErrorResponse,
+>(
+  provider: string,
+  customTransformer?: CustomTransformer<OpenAIChatCompleteResponse, T>
+) => {
+  const transformer: (
+    response: T | ErrorResponse,
+    responseStatus: number
+  ) => T | ErrorResponse = (response, responseStatus) => {
+    if (responseStatus !== 200 && 'error' in response) {
+      const errorResponse = OpenAIErrorResponseTransform(
+        response,
+        provider ?? OPEN_AI
+      );
+      if (customTransformer) {
+        return customTransformer(response as ErrorResponse, true);
+      }
+
+      return errorResponse;
+    }
+
+    if (customTransformer) {
+      return customTransformer(response as T);
+    }
+
+    Object.defineProperty(response, 'provider', {
+      value: provider,
+      enumerable: true,
+    });
+    return response;
+  };
+
+  return transformer;
+};
+
+/**
+ *
+ * @param provider Provider value
+ * @param options Enable transformer functions to specific task (complete, chatComplete or embed)
+ * @returns
+ */
+export const responseTransformers = <
+  T extends EmbedResponse | ErrorResponse,
+  U extends OpenAICompleteResponse | ErrorResponse,
+  V extends OpenAIChatCompleteResponse | ErrorResponse,
+>(
+  provider: string,
+  options: {
+    embed?: boolean | CustomTransformer<EmbedResponse | ErrorResponse, T>;
+    complete?:
+      | boolean
+      | CustomTransformer<OpenAICompleteResponse | ErrorResponse, U>;
+    chatComplete?:
+      | boolean
+      | CustomTransformer<OpenAIChatCompleteResponse | ErrorResponse, V>;
+  }
+) => {
+  const transformers: Record<string, Function | null> = {
+    complete: null,
+    chatComplete: null,
+    embed: null,
+  };
+
+  if (options.embed) {
+    transformers.embed = EmbedResponseTransformer<T>(provider);
+  }
+
+  if (options.complete) {
+    transformers.complete = CompleteResponseTransformer<U>(provider);
+  }
+
+  if (options.chatComplete) {
+    transformers.chatComplete = ChatCompleteResponseTransformer<V>(
+      provider,
+      typeof options.chatComplete === 'function'
```

Comment:
I think we can have the `function` type check for all other transformers also, right?

Reply:
Yep, they're not extended. Forget to extend them would do the needful.